graphs/gen/duplication.go

```diff
@@ -0,0 +1,121 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gen
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+
+	"github.com/gonum/graph"
+	"github.com/gonum/graph/simple"
+)
+
+// UndirectedMutator is an undirected graph builder that can remove edges.
+type UndirectedMutator interface {
+	graph.UndirectedBuilder
+	graph.EdgeRemover
+}
+
+// Duplication constructs a graph in the destination, dst, of order n. New nodes
+// are created by duplicating an existing node and all its edges. Each new edge is
+// deleted with probability delta. Additional edges are added between the new node
+// and existing nodes with probability alpha/|V|. An exception to this addition
+// rule is made for the parent node when sigma is not NaN; in this case an edge is
+// created with probability sigma. With the exception of the sigma parameter, this
+// corresponds to the completely correlated case in doi:10.1016/S0022-5193(03)00028-6.
+// If src is not nil it is used as the random source, otherwise rand.Float64 is used.
+func Duplication(dst UndirectedMutator, n int, sigma, alpha, delta float64, src *rand.Rand) error {
+	// As described in doi:10.1016/S0022-5193(03)00028-6 but
```

Comment:
For the first time I got to read a paper from the Journal of Theoretical Biology :-)